In [1]:
import seaborn as sns
data = sns.load_dataset('tips')

X = data[['total_bill']]
y = data['tip']

In [15]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [2]:
# Дополните код для работы программы на месте троеточия "..."

# Импортируйте функцию train_test_split и линейную регрессию
# А также MSE
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Разделите данные на тренировочную и тестовые выборки, с отношением 20%

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=38, test_size=0.2, shuffle=True)


In [4]:
# Инициализируйте и обучите модель на тренировочной выборке
model = LinearRegression()

model.fit(X_train,y_train)

LinearRegression()

In [7]:
# Сделайте предсказание на тестовой выборке
pred = model.predict(X_test)

# Выведите среднеквадратичную ошибку предсказания
print(mean_squared_error(pred, y_test))

0.7483593985916247


In [19]:
print(pred)

[2.51703018 2.33284418 2.05396369 4.62112105 2.6772824  2.12160261
 2.15802357 1.73866224 3.92287922 5.43799115 2.32139759 1.8645747
 2.06020728 3.47750291 1.70640367 2.30474801 3.57531921 4.26627684
 3.11225272 3.45565034 3.07999415 1.84480332 4.91352933 3.74909921
 3.01339583 2.72827174 2.2277437  3.28499213 2.34845316 2.25688046
 2.59299389 4.90728573 2.5846691  2.1330492  3.88333647 2.27977364
 2.84065642 4.05815707 2.45875664 4.5763753  2.7688551  2.16842956
 4.55556332 4.32350978 2.55032934 3.29227632 5.97494015 3.61174017
 3.08415655]


In [8]:
# Импортируйте кросс-валидацию
from sklearn.model_selection import cross_validate

# Инициализируйте кросс-валидацию с 5ю фолдами
model2 = LinearRegression()
cv = cross_validate(model2, X, y, cv = 5)


In [9]:
# Выведите результаты кросс-валидации и среднее
print(cv['test_score'])

print("Среднее отколнение:", cv['test_score'].mean())

[0.55056043 0.6192854  0.47423732 0.23796202 0.40067601]
Среднее отколнение: 0.45654423588476317


In [18]:
print(cv['score_time'])

[0.00298285 0.00145769 0.00141978 0.00144815 0.00150228]


Мы получили результаты используя инструменты рекомендованные в этом задании. Не совсем понятно как интерпретировать полученные результаты, так как не понятно по какому принципу оценивается результат 'test_score' для импортированного модуля cross_validate. Выше мы вычисляли среднеквадратичную ошибку предсказания и для тех же данных и той же модели линейной регресии мы получили гораздо более лучшие результаты  0.7483 против Среднее отколнение: 0.4565 для cross_validate.
Для лучшего понимания такого расхождения нужно изучить спецификацию модулей sklearn.